src/GovUk.Education.ExploreEducationStatistics.Data.Api/Services/PermalinkCsvMetaService.cs

```diff
@@ -0,0 +1,271 @@
+using GovUk.Education.ExploreEducationStatistics.Data.Api.Models;
+using GovUk.Education.ExploreEducationStatistics.Data.Api.Services.Interfaces;
+using GovUk.Education.ExploreEducationStatistics.Data.Api.ViewModels;
+using GovUk.Education.ExploreEducationStatistics.Data.Model;
+using GovUk.Education.ExploreEducationStatistics.Data.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Data.Services.Interfaces;
+using GovUk.Education.ExploreEducationStatistics.Data.Services.ViewModels.Meta;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+
+namespace GovUk.Education.ExploreEducationStatistics.Data.Api.Services;
+
+public class PermalinkCsvMetaService : IPermalinkCsvMetaService
+{
+    private readonly ILogger<PermalinkCsvMetaService> _logger;
+    private readonly ContentDbContext _contentDbContext;
+    private readonly StatisticsDbContext _statisticsDbContext;
+    private readonly IReleaseSubjectService _releaseSubjectService;
+    private readonly IReleaseFileBlobService _releaseFileBlobService;
+
+    public PermalinkCsvMetaService(
+        ILogger<PermalinkCsvMetaService> logger,
+        ContentDbContext contentDbContext,
+        StatisticsDbContext statisticsDbContext,
+        IReleaseSubjectService releaseSubjectService,
+        IReleaseFileBlobService releaseFileBlobService)
+    {
+        _logger = logger;
+        _contentDbContext = contentDbContext;
+        _statisticsDbContext = statisticsDbContext;
+        _releaseSubjectService = releaseSubjectService;
+        _releaseFileBlobService = releaseFileBlobService;
+    }
+
+    public async Task<Either<ActionResult, PermalinkCsvMetaViewModel>> GetCsvMeta(
+        LegacyPermalink permalink,
+        CancellationToken cancellationToken = default)
+    {
+        var releaseSubject = await _releaseSubjectService
+            .FindForLatestPublishedVersion(subjectId: permalink.Query.SubjectId);
+
+        var csvStream = releaseSubject is not null
+            ? await GetCsvStream(releaseSubject, cancellationToken)
+            : null;
+
+        var subjectMeta = permalink.FullTable.SubjectMeta;
+
+        var locations = await GetLocations(subjectMeta.LocationsHierarchical);
+
+        var filters = subjectMeta.Filters.Values.ToDictionary(
+            filter => filter.Name,
+            filter => new FilterCsvMetaViewModel(filter)
+        );
+        var indicators = subjectMeta.Indicators.ToDictionary(
+            indicator => indicator.Name,
+            indicator => new IndicatorCsvMetaViewModel(indicator)
+        );
+
+        var headers = await ListCsvHeaders(csvStream, locations, filters, indicators);
+
+        return new PermalinkCsvMetaViewModel
+        {
+            Locations = locations,
+            Filters = filters,
+            Indicators = indicators,
+            Headers = headers
+        };
+    }
+
+    private async Task<Stream?> GetCsvStream(
```

Comment:
They're essentially the same yes. For now, I'm happy to leave it as-is as we don't have a third case, so I'm going to err on the side of YAGNI.

Also want to avoid any more significant re-working of this PR! :weary: